### PR TITLE
[WIP][don't review] kubeadm: add Timeout structure and move TimeoutForControlPlane from APIServer into it

### LINKED ISSUE #2463: add API support for controlling various timeouts
edit: neolit123

ticket repurposed for a general / broader timeout support.

design draft :
https://docs.google.com/document/d/1q0OLHSD6M0JPjN8PxgvpJX1726omQh_Qd3os4LQqrtI/edit

---------

## What keywords did you search in kubeadm issues before filing this one?

["etcd timeout"](https://github.com/kubernetes/kubeadm/issues?q=is%3Aissue+etcd+timeout+) (5 open, 173 closed)

Similar questions, but not formal asks:

[#1712 (comment)](https://github.com/kubernetes/kubeadm/issues/1712#issuecomment-549893600)
[#2092 (comment by @neolit123)](https://github.com/kubernetes/kubeadm/issues/2092#issuecomment-607994592)

## Is this a BUG REPORT or FEATURE REQUEST?

FEATURE REQUEST
/kind feature
/kind api-change
/area etcd

Detailed motivation is covered in #2457,  but the short version is AWS would like to add support to CAPI for [Bottlerocket](https://github.com/bottlerocket-os/bottlerocket), and we have constraints that prevent us from just running `kubeadm join` without stopping part way along, running OS-specific configuration, and then resuming with `kubeadm join --skip-phases ...`. 

When bringing up etcd on additional control plane nodes, the subphase `control-plane-join/etcd`, kubeadm waits up to [40 seconds](https://github.com/kubernetes/kubernetes/blob/dd6ee9/cmd/kubeadm/app/phases/etcd/local.go#L208) for etcd to be available before exiting. In our case, we expect kubeadm to fail, and want to perform our own configuration when it does. We'd like to make this timeout configurable so that we can exit as soon as configuration and certificates are ready. Other users have asked [to increase this timeout](https://github.com/kubernetes/kubeadm/issues/1712#issuecomment-549893600)

I think a new `joinTimeout` option under `ClusterConfiguration.etcd.local` makes the most sense, but feel free to point me to a better location. (Similar to the existing `ClusterConfiguration.apiServer.timeoutForControlPlane` option)

## Versions

**kubeadm version** (use `kubeadm version`): All

**Environment**:
- **Kubernetes version** (use `kubectl version`): All
- **Cloud provider or hardware configuration**: AWS
- **OS** (e.g. from /etc/os-release): Bottlerocket
- **Kernel** (e.g. `uname -a`):
- **Others**:

cc @bcressey @jaxesn @vignesh-goutham

### PR DESCRIPTION
**Pull request by wangyysde**
…rHealthCheck into Timeout, drop TimeoutForControlPlane from APIServer

Signed-off-by: wangyysde <net_use@bzhy.com>

#### What type of PR is this?
/kind feature
/kind design

#### What this PR does / why we need it:
As   kubernetes/kubeadm#2463  discussion, we should have "timeouts" sub-structure under InitConfiguration and JoinConfiguration, that holds a number of different timeouts.
This PR add "timeouts" sub-structure under InitConfiguration and JoinConfiguration, and move TimeoutForControlPlane from APIServer into it.

#### Which issue(s) this PR fixes:

Fixes: kubernetes/kubeadm#2463

#### Special notes for your reviewer:

NONE

#### Does this PR introduce a user-facing change?

```release-note
TimeoutForControlPlane  removed from ClusterConfiguration.apiServer and using InitConfiguration/JoinConfiguration.timeout.apiServerHealthCheck instead of ClusterConfiguration.apiServer.timeoutForControlPlane 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

**Comment by pacoxu** [comment]
/retitle kubeadm: add Timeout structure and move TimeoutForControlPlane from APIServer into it
UT(TestRoundTripTypes) needs to be updated as well.
/priority important-soon
/triage-accepted

**Comment by wangyysde** [comment]
@pacoxu  Thanks. I have fix fuzzy test failed.

**Comment by MadhavJivrajani** [comment]
/remove-kind design
/kind feature

`kind/design` will soon be removed from `k/k` in favor of `kind/feature`. Relevant discussion can be found here: https://github.com/kubernetes/community/issues/5641

File: cmd/kubeadm/app/apis/kubeadm/types.go [hunk 1]
```diff
@@ -66,6 +66,10 @@ type InitConfiguration struct {
 	// Patches contains options related to applying patches to components deployed by kubeadm during
 	// "kubeadm init".
 	Patches *Patches
+
+	// Timeout holds timeout fields that relate to registering a new control-plane or node to the cluster,
+	// either via "kubeadm init" or "kubeadm join"
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/types.go:71]:
```suggestion
	// Timeouts holds fields controlling timeouts when registering a control plane node during
	// "kubeadm init".
```

similar for other places,

replace "init" -> "join" for JoinTimeouts.

File: cmd/kubeadm/app/apis/kubeadm/types.go [hunk 2]
```diff
@@ -200,6 +201,26 @@ type APIEndpoint struct {
 	BindPort int32
 }
 
+// InitTimeouts holds timeout fields that relate to registering a new control-plane to the cluster,
+// via "kubeadm init"
+type InitTimeouts struct {
+
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/types.go:207]:
```suggestion
```

File: cmd/kubeadm/app/apis/kubeadm/types.go [hunk 3]
```diff
@@ -200,6 +201,26 @@ type APIEndpoint struct {
...
+}
+
+// JoinTimeouts holds timeout fields that relate to registering a new control-plane or node to the cluster,
+// via "kubeadm join
+type JoinTimeouts struct {
+	// ApiServerHealthCheck controls the timeout that we use for API server to appear
+	// +optional
+	ApiServerHealthCheck *metav1.Duration `json:"apiServerHealthCheck,omitempty"`
+
+	// kubeletTLSBootstrap specifies how long kubeadm should wait for the kubelet to perform the TLS Bootstrap
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/types.go:220]:
```suggestion
	// kubeletTLSBootstrap specifies how long kubeadm should wait for the kubelet to perform TLS bootstrap.
	// +optional
```

File: cmd/kubeadm/app/apis/kubeadm/types.go [hunk 4]
```diff
@@ -200,6 +201,26 @@ type APIEndpoint struct {
...
+
+	// ApiServerHealthCheck controls the timeout that we use for API server to appear
+	// +optional
+	ApiServerHealthCheck *metav1.Duration `json:"apiServerHealthCheck,omitempty"`
+}
+
+// JoinTimeouts holds timeout fields that relate to registering a new control-plane or node to the cluster,
+// via "kubeadm join
+type JoinTimeouts struct {
+	// ApiServerHealthCheck controls the timeout that we use for API server to appear
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/types.go:216]:
```suggestion
	// ApiServerHealthCheck specifies how long kubeadm should check for the kube-apiserver to return
	// status OK on the /healthz endpoint during "kubeadm join".
```

File: cmd/kubeadm/app/apis/kubeadm/v1beta3/doc.go [hunk 5]
```diff
@@ -36,6 +36,17 @@ limitations under the License.
...
+// via "kubeadm init"
+// - Add JoinTimeouts to hold timeout fields that relate to registering a new control-plane or node to the cluster,
+// via "kubeadm join"
+// - Add Timeouts(InitTimeouts) field into InitConfiguration struct to holds timeout fields that relate to registering a new control-plane
+// to the cluster
+// - Add Timeouts(JoinTimeouts) field into JoinConfiguration struct to holds timeout fields that relate to registering a new control-plane or
+// node to the cluster
+// - Remove TimeoutForControlPlane field from APIServer, because TimeoutForControlPlane has been removed into InitTimeouts and JoinTimeouts
+// structure and renamed to ApiServerHealthCheck
+// - Add KubeletTLSBootstrap field into JoinTimeouts to specifies how long kubeadm should wait for the kubelet to perform the TLS Bootstrap
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/v1beta3/doc.go:49]: i think we should have less entries here:

e.g.
```
// - Add InitTimeouts to .... that contain A, B, C...
// - Add JoinTimeouts to .... that contain A, B, C...
```

File: cmd/kubeadm/app/apis/kubeadm/v1beta3/types.go [hunk 6]
```diff
@@ -200,6 +200,26 @@ type APIEndpoint struct {
 	BindPort int32 `json:"bindPort,omitempty"`
 }
 
+// InitTimeouts holds timeout fields that relate to registering a new control-planeto the cluster,
+// via "kubeadm init"
+type InitTimeouts struct {
+
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/v1beta3/types.go:206]:
```suggestion
```

File: cmd/kubeadm/app/apis/kubeadm/types.go [hunk 7]
```diff
@@ -200,6 +201,26 @@ type APIEndpoint struct {
...
+	// +optional
+	ApiServerHealthCheck *metav1.Duration `json:"apiServerHealthCheck,omitempty"`
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/types.go:218]: internal types (kubeadm/types.go) should not have:
- JSON tags
- +optional

File: cmd/kubeadm/app/apis/kubeadm/types.go [hunk 8]
```diff
@@ -342,6 +363,10 @@ type JoinConfiguration struct {
 	// Patches contains options related to applying patches to components deployed by kubeadm during
 	// "kubeadm join".
 	Patches *Patches
+
+	// Timeout holds timeout fields that relate to registering a new control-plane or node to the cluster,
+	// either via "kubeadm init" or "kubeadm join"
+	Timeouts JoinTimeouts `json:"timeout,omitempty"`
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/types.go:369]:
```suggestion
	Timeouts *JoinTimeouts `json:"timeout,omitempty"`
```

File: cmd/kubeadm/app/apis/kubeadm/types.go [hunk 9]
```diff
@@ -66,6 +66,10 @@ type InitConfiguration struct {
 	// Patches contains options related to applying patches to components deployed by kubeadm during
 	// "kubeadm init".
 	Patches *Patches
+
+	// Timeout holds timeout fields that relate to registering a new control-plane or node to the cluster,
+	// either via "kubeadm init" or "kubeadm join"
+	Timeouts InitTimeouts `json:"timeout,omitempty"`
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/types.go:72]:
```suggestion
	Timeouts *InitTimeouts `json:"timeout,omitempty"`
```

Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/types.go:72]: should be a pointer ideally.

File: cmd/kubeadm/app/apis/kubeadm/v1beta3/types.go [hunk 10]
```diff
@@ -65,6 +65,10 @@ type InitConfiguration struct {
 	// "kubeadm init".
 	// +optional
 	Patches *Patches `json:"patches,omitempty"`
+
+	// Timeout holds timeout fields that relate to registering a new control-plane or node to the cluster,
+	// either via "kubeadm init" or "kubeadm join"
+	Timeouts InitTimeouts `json:"timeout,omitempty"`
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/v1beta3/types.go:71]:
```suggestion
	// +optional
	Timeouts *InitTimeouts `json:"timeout,omitempty"`
```

- should be a pointer.
- have +optional

File: cmd/kubeadm/app/apis/kubeadm/v1beta3/types.go [hunk 11]
```diff
@@ -368,6 +388,10 @@ type JoinConfiguration struct {
 	// "kubeadm join".
 	// +optional
 	Patches *Patches `json:"patches,omitempty"`
+
+	// Timeout holds timeout fields that relate to registering a new control-plane or node to the cluster,
+	// either via "kubeadm init" or "kubeadm join"
+	Timeouts JoinTimeouts `json:"timeout,omitempty"`
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/v1beta3/types.go:394]:
```suggestion
	// either via "kubeadm init" or "kubeadm join"
	// +optional
	Timeouts *JoinTimeouts `json:"timeout,omitempty"`
```

File: cmd/kubeadm/app/cmd/phases/init/waitcontrolplane.go [hunk 12]
```diff
@@ -91,7 +91,15 @@ func runWaitControlPlanePhase(c workflow.RunData) error {
...
-	timeout := data.Cfg().ClusterConfiguration.APIServer.TimeoutForControlPlane.Duration
+	// Because Timeouts and ApiServerHealthCheck are not exist in v1beta2
+	// move set default value of ApiServerHealthCheck from defaults.go to here
+	// otherwise fuzzer and round trip test fail
+	var timeout time.Duration
+	if data.Cfg().Timeouts.ApiServerHealthCheck == nil {
+		timeout = kubeadmconstants.DefaultControlPlaneTimeout
+	} else {
+		timeout = data.Cfg().Timeouts.ApiServerHealthCheck.Duration
+	}
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/init/waitcontrolplane.go:102]: as we discussed on the other PR.
defaulting should not happen on runtime for static values that don't depend on host conditions.

this defaulting should be done in v1beta3/defaults.go

File: cmd/kubeadm/app/cmd/phases/join/kubelet.go [hunk 13]
```diff
@@ -170,11 +171,20 @@ func runKubeletStartJoinPhase(c workflow.RunData) (returnErr error) {
...
 
+	// if KubeletTLSBootstrap not specified then set KubeletTLSBootstrap to default value
+	// otherwise set it to cfg.Timeouts.KubeletTLSBootstrap
+	var timeout time.Duration
+	if cfg.Timeouts.KubeletTLSBootstrap == nil {
+		timeout = kubeadmconstants.KubeletTLSBootstrap
+	} else {
+		timeout = cfg.Timeouts.KubeletTLSBootstrap.Duration
+	}
+
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/join/kubelet.go:182]: [1]

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 14]
```diff
@@ -41,14 +42,23 @@ import (
...
 	// Get the required hostpath mounts
 	mounts := getHostPathVolumesForTheControlPlane(cfg)
+	// Because Timeouts and ApiServerHealthCheck are not exist in v1beta2
+	// move set default value of ApiServerHealthCheck from defaults.go to here
+	// otherwise fuzzer and round trip test fail
+	if initCfg.Timeouts.ApiServerHealthCheck == nil {
+		initCfg.Timeouts.ApiServerHealthCheck = &metav1.Duration{
+			Duration: kubeadmconstants.DefaultControlPlaneTimeout,
+		}
+	}
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/controlplane/manifests.go:61]: [1]

File: cmd/kubeadm/app/cmd/phases/init/controlplane.go [hunk 15]
```diff
@@ -148,6 +148,6 @@ func runControlPlaneSubphase(component string) func(c workflow.RunData) error {
 		cfg := data.Cfg()
 
 		fmt.Printf("[control-plane] Creating static Pod manifest for %q\n", component)
-		return controlplane.CreateStaticPodFiles(data.ManifestDir(), data.PatchesDir(), &cfg.ClusterConfiguration, &cfg.LocalAPIEndpoint, data.DryRun(), component)
+		return controlplane.CreateStaticPodFiles(data.ManifestDir(), data.PatchesDir(), cfg, &cfg.LocalAPIEndpoint, data.DryRun(), component)
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/init/controlplane.go:151]: i think we should pass separate arguments for timeout.
[2]

File: cmd/kubeadm/app/cmd/phases/init/etcd.go [hunk 16]
```diff
@@ -94,7 +94,7 @@ func runEtcdPhaseLocal() func(c workflow.RunData) error {
 				fmt.Printf("[dryrun] Would ensure that %q directory is present\n", cfg.Etcd.Local.DataDir)
 			}
 			fmt.Printf("[etcd] Creating static Pod manifest for local etcd in %q\n", data.ManifestDir())
-			if err := etcdphase.CreateLocalEtcdStaticPodManifestFile(data.ManifestDir(), data.PatchesDir(), cfg.NodeRegistration.Name, &cfg.ClusterConfiguration, &cfg.LocalAPIEndpoint, data.DryRun()); err != nil {
+			if err := etcdphase.CreateLocalEtcdStaticPodManifestFile(data.ManifestDir(), data.PatchesDir(), cfg.NodeRegistration.Name, cfg, &cfg.LocalAPIEndpoint, data.DryRun()); err != nil {
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/init/etcd.go:97]: [2]

File: cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go [hunk 17]
```diff
@@ -148,7 +148,7 @@ func runEtcdPhase(c workflow.RunData) error {
 	// etcdctl member add informs the cluster about the new member and the new member successfully establishing a connection to the
 	// existing one."
 	// TODO: add support for join dry-run: https://github.com/kubernetes/kubeadm/issues/2505
-	if err := etcdphase.CreateStackedEtcdStaticPodManifestFile(client, kubeadmconstants.GetStaticPodDirectory(), data.PatchesDir(), cfg.NodeRegistration.Name, &cfg.ClusterConfiguration, &cfg.LocalAPIEndpoint, false /* isDryRun */); err != nil {
+	if err := etcdphase.CreateStackedEtcdStaticPodManifestFile(client, kubeadmconstants.GetStaticPodDirectory(), data.PatchesDir(), cfg.NodeRegistration.Name, cfg, &cfg.LocalAPIEndpoint, false /* isDryRun */); err != nil {
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go:151]: [2]

File: cmd/kubeadm/app/cmd/phases/join/controlplaneprepare.go [hunk 18]
```diff
@@ -194,7 +194,7 @@ func runControlPlanePrepareControlPlaneSubphase(c workflow.RunData) error {
 		err := controlplane.CreateStaticPodFiles(
 			kubeadmconstants.GetStaticPodDirectory(),
 			data.PatchesDir(),
-			&cfg.ClusterConfiguration,
+			cfg,
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/join/controlplaneprepare.go:197]: [2]

File: cmd/kubeadm/app/constants/constants.go [hunk 19]
```diff
@@ -180,8 +180,8 @@ const (
 	DiscoveryRetryInterval = 5 * time.Second
 	// PatchNodeTimeout specifies how long kubeadm should wait for applying the label and taint on the control-plane before timing out
 	PatchNodeTimeout = 2 * time.Minute
-	// TLSBootstrapTimeout specifies how long kubeadm should wait for the kubelet to perform the TLS Bootstrap
-	TLSBootstrapTimeout = 5 * time.Minute
+	// kubeletTLSBootstrap specifies how long kubeadm should wait for the kubelet to perform the TLS Bootstrap
+	KubeletTLSBootstrap = 5 * time.Minute
```
Comment (neolit123) [inline cmd/kubeadm/app/constants/constants.go:184]: is this rename needed?

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 20]
```diff
@@ -41,14 +42,23 @@ import (
 // CreateInitStaticPodManifestFiles will write all static pod manifest files needed to bring up the control plane.
 func CreateInitStaticPodManifestFiles(manifestDir, patchesDir string, cfg *kubeadmapi.InitConfiguration, isDryRun bool) error {
 	klog.V(1).Infoln("[control-plane] creating static Pod files")
-	return CreateStaticPodFiles(manifestDir, patchesDir, &cfg.ClusterConfiguration, &cfg.LocalAPIEndpoint, isDryRun, kubeadmconstants.KubeAPIServer, kubeadmconstants.KubeControllerManager, kubeadmconstants.KubeScheduler)
+	return CreateStaticPodFiles(manifestDir, patchesDir, cfg, &cfg.LocalAPIEndpoint, isDryRun, kubeadmconstants.KubeAPIServer, kubeadmconstants.KubeControllerManager, kubeadmconstants.KubeScheduler)
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/controlplane/manifests.go:45]: [2]

**Review by neolit123 — COMMENTED** [review]
thanks for the PR @wangyysde this seems to be in the right direction.
i've added number of comments.

given code freeze is Thursday next week, i'm not sure how much time i will have to review this again until then. if we can't get it for v1beta3, it has to be scheduled for v1beta4.

File: cmd/kubeadm/app/apis/kubeadm/types.go (hunk 7 above)
Comment (wangyysde) [reply cmd/kubeadm/app/apis/kubeadm/types.go:218]: OK, Thanks.

File: cmd/kubeadm/app/cmd/phases/init/waitcontrolplane.go (hunk 12 above)
Comment (wangyysde) [reply cmd/kubeadm/app/cmd/phases/init/waitcontrolplane.go:102]: En, Thanks.

File: cmd/kubeadm/app/constants/constants.go (hunk 19 above)
Comment (wangyysde) [reply cmd/kubeadm/app/constants/constants.go:184]: It is not need to be renamed. But we try to add "KubeletTLSBootstrap" to set  the kubelet bootstrap timeout be configurable, and default value of KubeletTLSBootstrap is come from TLSBootstrapTimeout . So I try to rename TLSBootstrapTimeout  to KubeletTLSBootstrap  for conformance.

**Comment by wangyysde** [comment]
@neolit123 I have prepared a document at here: https://docs.google.com/document/d/1q0OLHSD6M0JPjN8PxgvpJX1726omQh_Qd3os4LQqrtI
